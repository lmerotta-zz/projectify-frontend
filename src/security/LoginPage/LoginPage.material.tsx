import {
  Grid,
  Typography,
  Box,
  TextField,
  Link,
  Button,
  Divider,
} from "@mui/material";
import { GitHub } from "@mui/icons-material";
import { grey } from "@mui/material/colors";
import background from "../images/background.svg";

const LoginPage = () => {
  return (
    <Grid container sx={{ minHeight: "100vh" }} spacing={0}>
      <Grid
        item
        xs={12}
        container
        direction="column"
        justifyContent="center"
        flexShrink={1}
        py={7}
        px={4}
        sx={{ backgroundImage: `url(${background})`, backgroundSize: "cover" }}
      >
        <Box display="flex" justifyContent="center" mb={6}>
          <Typography
            variant="h4"
            component="h2"
            color="white"
            sx={{ fontWeight: "bold" }}
          >
            ProjectifyTmp
          </Typography>
        </Box>
        <Typography
          color="white"
          variant="h6"
          component="h4"
          mb={1}
          sx={{ fontWeight: "normal" }}
        >
          Your{" "}
          <Typography color="secondary" component="span" variant="h6">
            projects
          </Typography>
          , your way
        </Typography>
        <Typography
          color="white"
          variant="body2"
          sx={{ fontWeight: "lighter" }}
        >
          Lorem ipsum dolor sit amet, consectetur adipiscing elit. Suspendisse
          quis nunc a mauris faucibus aliquet. Nullam sed
        </Typography>
      </Grid>
      <Grid item xs={12} flexGrow={1} p={4}>
        <Typography variant="h4" component="h1" sx={{ fontWeight: "medium" }}>
          Login
        </Typography>
        <Typography
          variant="subtitle1"
          component="h3"
          color={grey[500]}
          sx={{ fontWeight: "medium" }}
        >
          Log in to access your projects
        </Typography>
        <Box component="form" my={6}>
          <Grid container spacing={3}>
            <Grid item xs={12}>
              <TextField id="email" label="Email" fullWidth />
            </Grid>
            <Grid item xs={12}>
              <TextField
                id="password"
                type="password"
                label="Password"
                fullWidth
              />
              <Link
                display="block"
                mt={2}
                href="#"
                underline="hover"
                variant="body2"
                color={grey[500]}
              >
                Forgot password ?
              </Link>
            </Grid>
            <Grid item xs={12}>
              <Button
                type="submit"
                color="primary"
                variant="contained"
                fullWidth
                size="large"
              >
                Log in
              </Button>
              <Typography
                mt={2}
                color={grey[500]}
                variant="body2"
                sx={{ fontWeight: "medium" }}
              >
                Don't have an account?{" "}
                <Link
                  href="#"
                  color="secondary"
                  underline="hover"
                  variant="inherit"
                >
                  Register here
                </Link>
              </Typography>
            </Grid>
            <Grid item xs={12}>
              <Divider>
                <Typography color={grey[500]}>Or</Typography>
              </Divider>
            </Grid>
            <Grid item xs={12}>
              <Button
                variant="outlined"
                color="inherit"
                size="large"
                startIcon={<GitHub />}
              >
                Login with GitHub
              </Button>
            </Grid>
          </Grid>
        </Box>
      </Grid>
    </Grid>
  );
};

export default LoginPage;
