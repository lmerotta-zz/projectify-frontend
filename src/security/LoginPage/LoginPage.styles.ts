import { Link } from "components";
import tw, { styled } from "twin.macro";

export const ForgotPasswordLink = styled(Link)`
  ${tw`font-light text-sm lg:text-xs`}
`;

export const Divider = styled.p`
  &:before,
  &:after {
    content: " ";
    height: 1px;
    ${tw`bg-gray-300 absolute top-1/2 -translate-y-1/2`}
  }

  &:before {
    ${tw`left-0 w-2/5`}
  }
  &:after {
    ${tw`right-0 w-2/5`}
  }

  ${tw`bg-white text-center relative my-5 text-default`}
`;

export const GithubLink = styled.a`
  ${tw`px-3 py-2 border border-gray-400 text-gray-600 transition duration-300 bg-white rounded-md inline-flex items-center hover:bg-gray-100 active:bg-gray-300`}

  & > img {
    ${tw`object-contain h-8 mr-2`}
  }
`;
