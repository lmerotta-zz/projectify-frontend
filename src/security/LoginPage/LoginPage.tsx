/** @jsxImportSource @emotion/react */

import { Button, Input, Link, SubTitle, Title } from "components";
import { FormProvider, useForm } from "react-hook-form";
import tw, { styled } from "twin.macro";
import background from "./images/background.svg";

const Wrapper = tw.div`flex flex-col lg:flex-row min-h-screen`;

const Hero = styled.div`
  ${tw`w-full px-8 py-16 flex flex-col items-center justify-center space-y-12 lg:w-2/3`}
  background-image: url(${background});
  background-size: cover;
`;

const HeroContent = tw.div`lg:self-start lg:w-2/3`;

const LoginPage = () => {
  const form = useForm();

  return (
    <Wrapper>
      <Hero>
        <Title tw="text-light lg:my-auto">ProjectifyTmp</Title>
        <HeroContent>
          <SubTitle tw="text-white font-normal mb-3 lg:text-6xl">
            Your <span tw="text-secondary font-medium">projects</span>, your
            way.
          </SubTitle>
          <p tw="text-white font-light text-sm lg:text-xl">
            Lorem ipsum dolor sit amet, consectetur adipiscing elit. Suspendisse
            quis nunc a mauris faucibus aliquet. Nullam sed{" "}
          </p>
        </HeroContent>
      </Hero>
      <div tw="w-full flex-1 bg-white flex flex-col px-8 py-10 lg:w-1/3 md:px-48 lg:px-16 md:justify-center">
        <Title>Login</Title>
        <SubTitle tw="mb-10 lg:mb-20">Log in to access your projects.</SubTitle>
        <FormProvider {...form}>
          <div tw="mb-5 lg:mb-8">
            <Input name="email" type="email" label="Email" />
          </div>
          <div tw="mb-8 lg:mb-2">
            <Input name="password" type="password" label="Password" />
          </div>
          <Link href="/pek" tw="font-light text-sm mb-10 lg:text-xs">
            Forgot password?
          </Link>

          <div tw="flex justify-evenly items-center flex-col">
            <Button type="submit" tw="w-full mb-5 flex-1 py-4 font-bold">
              Log In
            </Button>

            <span tw="text-sm text-default">
              Don't have an account ?{" "}
              <Link href="/pek" color="secondary">
                Register here
              </Link>
            </span>
          </div>
        </FormProvider>
      </div>
    </Wrapper>
  );
};

export default LoginPage;
