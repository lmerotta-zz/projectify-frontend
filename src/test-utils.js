import React from "react";
import { render } from "@testing-library/react";
import { CacheProvider } from "@emotion/react";
import createCache from "@emotion/cache";

const cache = createCache({ key: "iodajdoiwanoawdni" });
cache.compat = true;

const AllTheProviders = ({ children }) => {
  return <CacheProvider value={cache}>{children}</CacheProvider>;
};

const customRender = (ui, options) =>
  render(ui, { wrapper: AllTheProviders, ...options });

// re-export everything
export * from "@testing-library/react";

// override render method
export { customRender as render };
