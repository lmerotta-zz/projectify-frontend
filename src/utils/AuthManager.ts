import { UserManager } from "oidc-client";

class AuthManager {
  private userManager: UserManager;
  constructor() {
    this.userManager = new UserManager({
      authority: process.env.REACT_APP_ACTIONS_URL,
      response_type: "code",
      client_id: process.env.REACT_APP_OAUTH_CLIENTID,
      redirect_uri: process.env.REACT_APP_OAUTH_REDIRECTURI,
      silent_redirect_uri: process.env.REACT_APP_OAUTH_SILENT_REDIRECTURI,
      scope: "email",
    });
  }

  async login(args: any) {
    await this.userManager.signinRedirect(args);
  }

  async isLoggedIn() {
    return await this.userManager.signinSilent({
      scope: "email",
    });
  }

  async getUser() {
    return await this.userManager.getUser();
  }

  async removeUser() {
    await this.userManager.removeUser();
  }

  async logout() {
    await this.userManager.signoutRedirect();
  }
}

export default new AuthManager();
