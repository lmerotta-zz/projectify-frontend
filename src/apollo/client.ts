import {
  ApolloClient,
  InMemoryCache,
  from,
  createHttpLink,
} from "@apollo/client";
import { RestLink } from "apollo-link-rest";
import { setContext } from "@apollo/client/link/context";
import i18next from "i18next";
import AuthManager from "utils/AuthManager";
import { isAuthenticated } from "./local-state";

const restLink = new RestLink({
  endpoints: {
    login: {
      uri: `${process.env.REACT_APP_ACTIONS_URL}/login`,
    },
  },
  credentials: "include",
});

const httpLink = createHttpLink({
  uri: process.env.REACT_APP_GRAPHQL_URL,
});

const authLink = setContext(async (_, { headers }) => {
  if (!isAuthenticated()) {
    return;
  }
  let user = await AuthManager.getUser();
  if (user === null || user.expired) {
    try {
      user = await AuthManager.isLoggedIn();
    } catch (e) {
      user = null;
    }
  }
  if (user !== null) {
    return {
      headers: {
        ...headers,
        Authorization: `Bearer ${user.id_token}`,
      },
    };
  }

  return {
    headers,
  };
});

let locale = "";

i18next.on("languageChanged", (lng) => {
  locale = lng;
});

const client = new ApolloClient({
  cache: new InMemoryCache(),
  headers: {
    "x-locale": locale,
  },
  link: from([restLink, authLink.concat(httpLink)]),
});

export default client;
