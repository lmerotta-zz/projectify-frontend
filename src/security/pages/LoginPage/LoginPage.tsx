import { gql } from "@apollo/client";
import { yupResolver } from "@hookform/resolvers/yup";
import { GitHub } from "@mui/icons-material";
import { Button, Divider, Grid, Link, Typography } from "@mui/material";
import { TextField } from "components";
import { useLoginMutation } from "generated/graphql";
import { FormProvider, useForm } from "react-hook-form";
import { Trans, useTranslation } from "react-i18next";
import { useLocation } from "react-router";
import { Link as RouterLink } from "react-router-dom";
import { toast } from "react-toastify";
import { RightPane, RightPaneFormWrapper } from "security/components";
import AuthManager from "utils/AuthManager";
import mapViolationsToForm from "utils/mapViolationsToForm";
import * as yup from "yup";

export const LOGIN_MUTATION = gql`
  mutation login($username: String!, $password: String!) {
    login(input: { username: $username, password: $password })
      @rest(
        type: "Login"
        path: ""
        endpoint: "login"
        method: "POST"
        bodyKey: "input"
      ) {
      NoResponse
    }
  }
`;

type LoginFormType = {
  email: string;
  password: string;
};

const schema = yup.object().shape({
  email: yup.string().email().required(),
  password: yup.string().required(),
});

const LoginPage = () => {
  const form = useForm<LoginFormType>({
    resolver: yupResolver(schema),
  });
  const { state } = useLocation<{ referrer: Location }>();

  const referrer = state?.referrer
    ? `${state.referrer.pathname}${state.referrer.search}${state.referrer.hash}`
    : "/";

  const { t } = useTranslation();

  const [login] = useLoginMutation({
    onError: (e) => {
      /* istanbul ignore else */
      if (!mapViolationsToForm(form.setError, e)) {
        toast.error(t("global.errors.internal-server-error"));
      }
    },
    onCompleted: async () => {
      await AuthManager.login({ state: referrer });
    },
  });

  return (
    <RightPane>
      <Typography variant="h4" component="h1" sx={{ fontWeight: "medium" }}>
        {t("security.login_page.page_title")}
      </Typography>
      <Typography
        variant="subtitle1"
        component="h3"
        color="neutral"
        sx={{ fontWeight: "medium" }}
      >
        {t("security.login_page.page_subtitle")}
      </Typography>
      <FormProvider {...form}>
        <RightPaneFormWrapper
          component="form"
          onSubmit={form.handleSubmit(async (data) => {
            await login({
              variables: { username: data.email, password: data.password },
            });
          })}
        >
          <Grid container spacing={3}>
            <Grid item xs={12}>
              <TextField
                type="email"
                id="email"
                label={t("security.login_page.form.label_email")}
                fullWidth
                variant="standard"
                {...form.register("email")}
              />
            </Grid>
            <Grid item xs={12}>
              <TextField
                id="password"
                type="password"
                label={t("security.login_page.form.label_password")}
                fullWidth
                variant="standard"
                {...form.register("password")}
              />
              <Link
                display="block"
                component={RouterLink}
                mt={2}
                to="/pek"
                underline="hover"
                variant="body2"
                color="neutral"
              >
                {t("security.login_page.form.forgot_password_link")}
              </Link>
            </Grid>
            <Grid item xs={12}>
              <Button
                type="submit"
                color="primary"
                variant="contained"
                fullWidth
                size="large"
                disabled={form.formState.isSubmitting}
              >
                {t("security.login_page.form.btn_login")}
              </Button>
              <Typography mt={2} color="neutral" variant="body2">
                <Trans
                  i18nKey="security.login_page.form.register_link"
                  components={{
                    Link: (
                      <Link
                        component={RouterLink}
                        to="/security/register"
                        color="secondary"
                        underline="hover"
                        variant="inherit"
                      />
                    ),
                  }}
                />
              </Typography>
            </Grid>
            <Grid item xs={12}>
              <Divider>
                <Typography color="neutral">
                  {t("security.login_page.form.label_or")}
                </Typography>
              </Divider>
            </Grid>
            <Grid item xs={12}>
              <Button
                component={Link}
                variant="outlined"
                color="inherit"
                href={`${process.env.REACT_APP_ACTIONS_URL}/oauth/connect/github?_destination=${process.env.REACT_APP_ACTIONS_URL}/oauth/connected?target=${referrer}`}
                size="large"
                startIcon={<GitHub />}
              >
                {t("security.login_page.form.oauth_github_login")}
              </Button>
            </Grid>
          </Grid>
        </RightPaneFormWrapper>
      </FormProvider>
    </RightPane>
  );
};

export default LoginPage;
