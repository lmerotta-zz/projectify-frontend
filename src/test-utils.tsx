import { MockedProvider, MockedProviderProps } from "@apollo/client/testing";
import {
  CircularProgress,
  StyledEngineProvider,
  ThemeProvider,
} from "@mui/material";
import { render as realRender, RenderOptions } from "@testing-library/react";
import { ReactElement, ReactNode, Suspense } from "react";
import { MemoryRouter, MemoryRouterProps } from "react-router";
import { Route, Routes } from "react-router-dom";
import { ToastContainer } from "react-toastify";
import theme from "theme";

function renderWithProviders(
  ui: ReactElement,
  options?: Omit<RenderOptions, "queries"> & {
    routerProps?: MemoryRouterProps;
    graphqlProps?: MockedProviderProps;
  }
) {
  const Wrapper = ({ children }: { children?: ReactNode }) => {
    return (
      <StyledEngineProvider injectFirst>
        <ThemeProvider theme={theme}>
          <ToastContainer
            theme="colored"
            autoClose={3000}
            containerId="react-toastify"
          />
          <MockedProvider {...options?.graphqlProps} addTypename={false}>
            <MemoryRouter {...options?.routerProps}>
              <Suspense fallback={<CircularProgress />}>
                <Routes>
                  <Route
                    path="/*"
                    element={
                      <>
                        <h1>TESTSASATATAST</h1>
                        {children}
                      </>
                    }
                  />
                </Routes>
              </Suspense>
            </MemoryRouter>
          </MockedProvider>
        </ThemeProvider>
      </StyledEngineProvider>
    );
  };

  return realRender(ui, { wrapper: Wrapper, ...options });
}

export * from "@testing-library/react";
export { default as userEvent } from "@testing-library/user-event";
export { renderWithProviders };
