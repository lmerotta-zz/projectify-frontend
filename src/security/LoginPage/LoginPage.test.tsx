import { MockedProvider } from "@apollo/client/testing";
import { render, fireEvent, act, waitFor } from "test-utils";
import { MemoryRouter } from "react-router";
import AuthManager from "utils/AuthManager";
import LoginPage, { LOGIN_MUTATION } from "./LoginPage";
import { toast } from "react-toastify";

jest.mock("utils/AuthManager", () => ({
  login: jest.fn(),
}));

describe("LoginPage unit tests", () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it("Returns invalid credential on wrong username login", async () => {
    const toastSpy = jest.spyOn(toast, "error");

    const mocks = [
      {
        request: {
          query: LOGIN_MUTATION,
          variables: {
            username: "failed@test.com",
            password: "test",
          },
        },
        error: {
          result: {
            error: "test",
          },
        },
      },
    ];

    const result = render(
      <MockedProvider mocks={mocks as any} addTypename={false}>
        <MemoryRouter>
          <LoginPage />
        </MemoryRouter>
      </MockedProvider>
    );

    fireEvent.input(result.getByTestId("input-email"), {
      target: { value: "failed@test.com" },
    });
    fireEvent.input(result.getByTestId("input-password"), {
      target: { value: "test" },
    });

    await act(async () => {
      fireEvent.click(result.getByTestId("btn-login"));
    });

    await act(async () => {
      await new Promise((resolve) => setTimeout(resolve, 0));
    });

    await waitFor(() => expect(toastSpy).toHaveBeenCalled());
  });

  it("Stores the oauth code when request succeeds", async () => {
    const toastSpy = jest.spyOn(toast, "error");

    const mocks = [
      {
        request: {
          query: LOGIN_MUTATION,
          variables: {
            username: "success@test.com",
            password: "test",
          },
        },
        result: {
          data: null,
        },
      },
    ];

    const result = render(
      <MockedProvider mocks={mocks as any} addTypename={false}>
        <MemoryRouter>
          <LoginPage />
        </MemoryRouter>
      </MockedProvider>
    );

    expect(result.asFragment()).toMatchSnapshot();

    fireEvent.input(result.getByTestId("input-email"), {
      target: { value: "success@test.com" },
    });
    fireEvent.input(result.getByTestId("input-password"), {
      target: { value: "test" },
    });

    await act(async () => {
      fireEvent.click(result.getByTestId("btn-login"));
    });

    await act(async () => {
      await new Promise((resolve) => setTimeout(resolve, 1000));
    });

    expect(toastSpy).not.toHaveBeenCalled();
    expect(AuthManager.login).toHaveBeenCalled();
  });
});
