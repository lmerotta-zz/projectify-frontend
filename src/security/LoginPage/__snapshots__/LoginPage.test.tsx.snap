// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginPage unit tests Stores the oauth code when request succeeds 1`] = `
<DocumentFragment>
  .emotion-0 {
  width: 100%;
  -webkit-flex: 1 1 0%;
  -ms-flex: 1 1 0%;
  flex: 1 1 0%;
  --tw-bg-opacity: 1;
  background-color: rgba(255, 255, 255, var(--tw-bg-opacity));
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  padding-left: 2rem;
  padding-right: 2rem;
  padding-top: 2.5rem;
  padding-bottom: 2.5rem;
}

@media (min-width: 768px) {
  .emotion-0 {
    padding-left: 12rem;
    padding-right: 12rem;
    -webkit-box-pack: center;
    -ms-flex-pack: center;
    -webkit-justify-content: center;
    justify-content: center;
  }
}

@media (min-width: 1024px) {
  .emotion-0 {
    width: 33.333333%;
    padding-left: 4rem;
    padding-right: 4rem;
  }
}

.emotion-1 {
  font-size: 2.25rem;
  line-height: 2.5rem;
  --tw-text-opacity: 1;
  color: rgba(39, 39, 42, var(--tw-text-opacity));
  font-weight: 700;
}

@media (min-width: 768px) {
  .emotion-1 {
    font-size: 3rem;
    line-height: 1;
  }
}

.emotion-2 {
  font-size: 1.25rem;
  line-height: 1.75rem;
  --tw-text-opacity: 1;
  color: rgba(113, 113, 122, var(--tw-text-opacity));
  font-weight: 500;
  margin-bottom: 2.5rem;
}

@media (min-width: 768px) {
  .emotion-2 {
    font-size: 1.5rem;
    line-height: 2rem;
  }
}

@media (min-width: 1024px) {
  .emotion-2 {
    margin-bottom: 5rem;
  }
}

.emotion-3 {
  margin-bottom: 1.25rem;
}

@media (min-width: 1024px) {
  .emotion-3 {
    margin-bottom: 2rem;
  }
}

.emotion-4 {
  position: relative;
  outline: 2px solid transparent;
  outline-offset: 2px;
}

.emotion-5 {
  width: 100%;
  position: relative;
}

.emotion-6 {
  width: 100%;
  padding-top: 1rem;
  padding-bottom: 0.5rem;
  padding-left: 0.75rem;
  padding-right: 0.75rem;
  border-width: 1px;
  border-radius: 0.25rem;
  --tw-border-opacity: 1;
  border-color: rgba(156, 163, 175, var(--tw-border-opacity));
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  outline: 2px solid transparent;
  outline-offset: 2px;
  transition-property: background-color,border-color,color,fill,stroke;
  transition-timing-function: cubic-bezier(0.4, 0, 0.2, 1);
  transition-duration: 150ms;
}

.emotion-6:focus {
  --tw-border-opacity: 1;
  border-color: rgba(147, 51, 234, var(--tw-border-opacity));
}

.emotion-6:focus+.label {
  --tw-text-opacity: 1;
  color: rgba(147, 51, 234, var(--tw-text-opacity));
  top: 0px;
  --tw-translate-y: 0.125rem;
  font-size: 0.75rem;
  line-height: 1rem;
}

.emotion-6:not(:placeholder-shown) {
  --tw-border-opacity: 1;
  border-color: rgba(147, 51, 234, var(--tw-border-opacity));
}

.emotion-6:not(:placeholder-shown)+.label {
  --tw-text-opacity: 1;
  color: rgba(147, 51, 234, var(--tw-text-opacity));
  top: 0px;
  --tw-translate-y: 0.125rem;
  font-size: 0.75rem;
  line-height: 1rem;
}

.emotion-7 {
  --tw-translate-x: 0;
  --tw-translate-y: -50%;
  --tw-rotate: 0;
  --tw-skew-x: 0;
  --tw-skew-y: 0;
  --tw-scale-x: 1;
  --tw-scale-y: 1;
  -webkit-transform: translateX(var(--tw-translate-x)) translateY(var(--tw-translate-y)) rotate(var(--tw-rotate)) skewX(var(--tw-skew-x)) skewY(var(--tw-skew-y)) scaleX(var(--tw-scale-x)) scaleY(var(--tw-scale-y));
  -moz-transform: translateX(var(--tw-translate-x)) translateY(var(--tw-translate-y)) rotate(var(--tw-rotate)) skewX(var(--tw-skew-x)) skewY(var(--tw-skew-y)) scaleX(var(--tw-scale-x)) scaleY(var(--tw-scale-y));
  -ms-transform: translateX(var(--tw-translate-x)) translateY(var(--tw-translate-y)) rotate(var(--tw-rotate)) skewX(var(--tw-skew-x)) skewY(var(--tw-skew-y)) scaleX(var(--tw-scale-x)) scaleY(var(--tw-scale-y));
  transform: translateX(var(--tw-translate-x)) translateY(var(--tw-translate-y)) rotate(var(--tw-rotate)) skewX(var(--tw-skew-x)) skewY(var(--tw-skew-y)) scaleX(var(--tw-scale-x)) scaleY(var(--tw-scale-y));
  position: absolute;
  top: 50%;
  left: 0.75rem;
  --tw-text-opacity: 1;
  color: rgba(113, 113, 122, var(--tw-text-opacity));
  cursor: text;
  transition-property: all;
  transition-timing-function: cubic-bezier(0.4, 0, 0.2, 1);
  transition-duration: 150ms;
  pointer-events: none;
}

.emotion-8 {
  margin-bottom: 1.25rem;
  margin-bottom: 2rem;
}

@media (min-width: 1024px) {
  .emotion-8 {
    margin-bottom: 2rem;
  }
}

@media (min-width: 1024px) {
  .emotion-8 {
    margin-bottom: 0.5rem;
  }
}

.emotion-13 {
  transition-property: background-color,border-color,color,fill,stroke,opacity,box-shadow,transform;
  transition-timing-function: cubic-bezier(0.4, 0, 0.2, 1);
  transition-duration: 150ms;
  --tw-text-opacity: 1;
  color: rgba(39, 39, 42, var(--tw-text-opacity));
  font-weight: 300;
  font-size: 0.875rem;
  line-height: 1.25rem;
}

.emotion-13:hover {
  --tw-text-opacity: 1;
  color: rgba(113, 113, 122, var(--tw-text-opacity));
}

@media (min-width: 1024px) {
  .emotion-13 {
    font-size: 0.75rem;
    line-height: 1rem;
  }
}

.emotion-14 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: space-evenly;
  -ms-flex-pack: space-evenly;
  -webkit-justify-content: space-evenly;
  justify-content: space-evenly;
  text-align: center;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  margin-top: 0.75rem;
}

.emotion-14>*:first-child {
  margin-bottom: 1.25rem;
  -webkit-flex: 1 1 0%;
  -ms-flex: 1 1 0%;
  flex: 1 1 0%;
  padding-top: 1rem;
  padding-bottom: 1rem;
  font-weight: 700;
}

.emotion-15 {
  transition-property: background-color,border-color,color,fill,stroke,opacity,box-shadow,transform;
  transition-timing-function: cubic-bezier(0.4, 0, 0.2, 1);
  transition-duration: 150ms;
  padding-left: 0.75rem;
  padding-right: 0.75rem;
  padding-top: 0.5rem;
  padding-bottom: 0.5rem;
  border-radius: 0.375rem;
  --tw-text-opacity: 1;
  color: rgba(255, 255, 255, var(--tw-text-opacity));
  --tw-bg-opacity: 1;
  background-color: rgba(147, 51, 234, var(--tw-bg-opacity));
}

.emotion-15:hover {
  --tw-bg-opacity: 1;
  background-color: rgba(126, 34, 206, var(--tw-bg-opacity));
}

.emotion-16 {
  font-size: 0.875rem;
  line-height: 1.25rem;
  --tw-text-opacity: 1;
  color: rgba(113, 113, 122, var(--tw-text-opacity));
}

.emotion-17 {
  --tw-bg-opacity: 1;
  background-color: rgba(255, 255, 255, var(--tw-bg-opacity));
  text-align: center;
  position: relative;
  margin-top: 1.25rem;
  margin-bottom: 1.25rem;
  --tw-text-opacity: 1;
  color: rgba(113, 113, 122, var(--tw-text-opacity));
}

.emotion-17:before,
.emotion-17:after {
  content: " ";
  height: 1px;
  --tw-bg-opacity: 1;
  background-color: rgba(209, 213, 219, var(--tw-bg-opacity));
  position: absolute;
  top: 50%;
  --tw-translate-y: -50%;
}

.emotion-17:before {
  left: 0px;
  width: 40%;
}

.emotion-17:after {
  right: 0px;
  width: 40%;
}

.emotion-18 {
  transition-property: background-color,border-color,color,fill,stroke,opacity,box-shadow,transform;
  transition-timing-function: cubic-bezier(0.4, 0, 0.2, 1);
  transition-duration: 300ms;
  padding-left: 0.75rem;
  padding-right: 0.75rem;
  padding-top: 0.5rem;
  padding-bottom: 0.5rem;
  border-width: 1px;
  --tw-border-opacity: 1;
  border-color: rgba(156, 163, 175, var(--tw-border-opacity));
  --tw-text-opacity: 1;
  color: rgba(75, 85, 99, var(--tw-text-opacity));
  --tw-bg-opacity: 1;
  background-color: rgba(255, 255, 255, var(--tw-bg-opacity));
  border-radius: 0.375rem;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.emotion-18:hover {
  --tw-bg-opacity: 1;
  background-color: rgba(243, 244, 246, var(--tw-bg-opacity));
}

.emotion-18:active {
  --tw-bg-opacity: 1;
  background-color: rgba(209, 213, 219, var(--tw-bg-opacity));
}

.emotion-18>img {
  object-fit: contain;
  height: 2rem;
  margin-right: 0.5rem;
}

<div
    class="emotion-0"
    style="opacity: 0;"
  >
    <h3
      class="emotion-1"
    >
      security.login_page.page_title
    </h3>
    <h4
      class="emotion-2"
    >
      security.login_page.page_subtitle
    </h4>
    <form>
      <div
        class="emotion-3"
      >
        <div
          class="emotion-4"
        >
          <div
            class="emotion-5"
          >
            <input
              class="emotion-6"
              data-testid="input-email"
              name="email"
              placeholder=" "
              type="email"
            />
            <label
              class="label emotion-7"
              data-testid="label-email"
            >
              security.login_page.form.label_email
            </label>
          </div>
        </div>
      </div>
      <div
        class="emotion-8"
      >
        <div
          class="emotion-4"
        >
          <div
            class="emotion-5"
          >
            <input
              class="emotion-6"
              data-testid="input-password"
              name="password"
              placeholder=" "
              type="password"
            />
            <label
              class="label emotion-7"
              data-testid="label-password"
            >
              security.login_page.form.label_password
            </label>
          </div>
        </div>
      </div>
      <a
        class="emotion-13"
        href="/pek"
      >
        security.login_page.form.forgot_password_link
      </a>
      <div
        class="emotion-14"
      >
        <button
          class="emotion-15"
          data-testid="btn-login"
          type="submit"
        >
          security.login_page.form.btn_login
        </button>
        <span
          class="emotion-16"
        >
          security.login_page.form.register_link
        </span>
      </div>
    </form>
    <div>
      <p
        class="emotion-17"
      >
        security.login_page.form.label_or
      </p>
      <a
        class="emotion-18"
        href="http://localhost:8001/actions/security/authentication/oauth/connect/github?_destination=http://localhost:8001/actions/security/authentication/oauth/connected?target=/"
      >
        <img
          alt="GitHub Logo"
          src="gh-logo.png"
        />
        security.login_page.form.oauth_github_login
      </a>
    </div>
  </div>
</DocumentFragment>
`;
